website/static/js/pages/FileEditor.js

```diff
@@ -0,0 +1,176 @@
+        switch(self.throttledStatus()) {
+            case 'connected':
+                return 'Live editing mode';
+            case 'connecting':
+                return 'Attempting to connect';
+            case 'unsupported':
+                return 'Unsupported browser';
+            default:
+                return 'Unavailable: Live editing';
+        }
+    });
+
+    self.progressBar = ko.computed(function() {
+        switch(self.throttledStatus()) {
+            case 'connected':
+                return {
+                    class: 'progress-bar progress-bar-success',
+                    style: 'width: 100%'
+                };
+
+            case 'connecting':
+                return {
+                    class: 'progress-bar progress-bar-warning progress-bar-striped active',
+                    style: 'width: 100%'
+                };
+            default:
+                return {
+                    class: 'progress-bar progress-bar-danger',
+                    style: 'width: 100%'
+                };
+        }
+    });
+
+    self.modalTarget = ko.computed(function() {
+        switch(self.throttledStatus()) {
+            case 'connected':
+                return '#connectedModal';
+            case 'connecting':
+                return '#connectingModal';
+            case 'unsupported':
+                return '#unsupportedModal';
+            default:
+                return '#disconnectedModal';
+        }
+    });
+
+    self.filesDiffer = function(file1, file2) {
+        // Handle inconsistencies in newline notation
+        var clean1 = typeof file1 === 'string' ?
+            file1.replace(/(\r\n|\n|\r)/gm, '\n') : '';
+        var clean2 = typeof file2 === 'string' ?
+            file2.replace(/(\r\n|\n|\r)/gm, '\n') : '';
+
+        return clean1 !== clean2;
+    };
+
+    self.changed = function() {
+        return self.filesDiffer(self.initText(), self.currentText());
+    };
+
+    // Fetch initial file text
+    self.fetchData = function() {
+        var request = $.ajax({
+            type: 'GET',
+            url: url,
+            dataType: 'json'
+        });
+        request.done(function (response) {
+            self.initText(response.content);
+
+        });
+        request.fail(function (xhr, textStatus, error) {
+            $osf.growl('Error','The file content could not be loaded.');
+            Raven.captureMessage('Could not GET file contents.', {
+                url: url,
+                textStatus: textStatus,
+                error: error
+            });
+        });
+        return request;
+    };
+
+    // Revert to last saved version, even if draft is more recent
+    self.revertChanges = function() {
+        return self.fetchData().then(function(response) {
+            // Dirty check now covers last saved version
+            self.initText(response.content);
+            self.currentText(response.content);
+        });
+    };
+
+    $(window).on('beforeunload', function() {
```

Comment:
We use this pattern in a couple of places. @chrisseto uses this on the files page when uploads are in progress. It would be good to generalize this and move it to osfHelpers.js